### PR TITLE
[4.2] Use layouts in color field

### PR DESCRIPTION
**Pull request by laoneo**
Followup Pull Request for pr #36551 and issue #36548.

### Summary of Changes
Adds the layout filter for the color custom field.

Additionally it deprecates the control parameter in favor of the layout one in the color filed to be more inline with the other core form fields.

~This should be tested after #36551 is merged.~

### Testing Instructions
- Create a color custom field
- Change the form layout to slider or simple in the options tab
- Edit an article

### Actual result BEFORE applying this Pull Request
Color field is always displayed with the same result.

### Expected result AFTER applying this Pull Request
Color field is displayed with different layouts.

### CODE REVIEW

File: plugins/fields/color/params/color.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<form>
+	<fields name="params">
+		<fieldset name="basic">
+			<fieldset name="formoptions">
+				<field
+					name="form_layout"
+					type="list"
+					label="COM_FIELDS_FIELD_FORM_LAYOUT_LABEL"
```
Comment (Quy) [inline plugins/fields/color/params/color.xml:9]: Missing language string.

File: administrator/language/en-GB/plg_fields_color.ini [hunk 2]
```diff
@@ -5,4 +5,7 @@
 
 PLG_FIELDS_COLOR="Fields - Colour"
 PLG_FIELDS_COLOR_LABEL="Colour (%s)"
+PLG_FIELDS_COLOR_PARAMS_FORM_LAYOUT_SIMPLE="Simple"
+PLG_FIELDS_COLOR_PARAMS_FORM_LAYOUT_ADVANCED="Advanced"
```
Comment (Quy) [inline administrator/language/en-GB/plg_fields_color.ini]:
```suggestion
PLG_FIELDS_COLOR_PARAMS_FORM_LAYOUT_ADVANCED="Advanced"
PLG_FIELDS_COLOR_PARAMS_FORM_LAYOUT_SIMPLE="Simple"
```

**Comment by crystalenka** [comment]
I have tested this item :red_circle: unsuccessfully on 59b08bb72ee1204fbda9737ed4b875739f9d9250<br /><br />All the layout options work except for the slider. For the slider, it appears, but it doesn't show the hue as expected. I'll add a screenshot in the next comment.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36553">issues.joomla.org/tracker/joomla-cms/36553</a>.</sub>

**Comment by crystalenka** [comment]
Slider screenshot. There's no color behind it so it looks empty.
Screen Shot 2022-01-09 at 12 30 35



<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36553">issues.joomla.org/tracker/joomla-cms/36553</a>.</sub>

**Comment by laoneo** [comment]
Even as I didn't change something on the JS side, can you rebuild it with npm or download a new build from here? Because on my dev server it worked after a `npm ci`.

**Comment by Quy** [comment]
It is reproducible with the prebuilt package.

**Comment by Quy** [comment]
The linear-gradient values are different.

Fields:
`<input type="range" min="0" max="360" class="form-control color-slider" id="hue-slider" data-type="hue" style="background: rgba(0, 0, 0, 0) linear-gradient(90deg, rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255), rgb(255, 255, 255)) repeat scroll 0% 0%; appearance: none;">`

Administrator Template Styles:
`<input type="range" min="0" max="360" class="form-control color-slider" id="hue-slider" data-type="hue" style="background: rgba(0, 0, 0, 0) linear-gradient(90deg, rgb(83, 19, 19), rgb(83, 19, 19), rgb(83, 38, 19), rgb(83, 57, 19), rgb(83, 77, 19), rgb(70, 83, 19), rgb(51, 83, 19), rgb(32, 83, 19), rgb(19, 83, 25), rgb(19, 83, 45), rgb(19, 83, 64), rgb(19, 83, 83), rgb(19, 64, 83), rgb(19, 45, 83), rgb(19, 25, 83), rgb(32, 19, 83), rgb(51, 19, 83), rgb(70, 19, 83), rgb(83, 19, 77), rgb(83, 19, 57), rgb(83, 19, 38), rgb(83, 19, 19), rgb(83, 19, 19)) repeat scroll 0% 0%; appearance: none;">`

**Comment by laoneo** [comment]
Found the issue, can you guys test again? Ignore the conflict, will disappear when 4.2 got the commits from 4.1 merged.

**Comment by Quy** [comment]
Before making a selection:
color-slider-pre

After making a selection and saving:
color-slider-post

**Comment by laoneo** [comment]
The slider shows the colors based on the initial color. So when you change it the range changes. Honestly I'm not much into this, perhaps @dgrammatiko can shed here some light why this happens. It is the same when you give it a default color.

**Comment by brianteeman** [comment]
@laoneo I think you are missing what @quy is saying. 

image

File: libraries/src/Form/Field/ColorField.php [hunk 3]
```diff
@@ -203,7 +204,7 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
 		if ($return)
 		{
 			$this->colors     = (string) $this->element['colors'];
-			$this->control    = isset($this->element['control']) ? (string) $this->element['control'] : 'hue';
+			$this->control    = isset($this->element['control']) ? (string) $this->element['control'] : '';
```
Comment (dgrammatiko) [inline libraries/src/Form/Field/ColorField.php:207]:
```suggestion
			$this->control    = isset($this->element['control']) ? (string) $this->element['control'] : 'hue';
```

**Comment by laoneo** [comment]
Don't think so as the color range of the slider is based on the available color. So when you choose one, the range changes. So what I do not understand is how to have a fixed color range which works in all situations.

**Comment by dgrammatiko** [comment]
>  perhaps @dgrammatiko can shed here some light why this happens.

I need to debug it but it seems that the removal of the `$control` is not right. In the advanced this var controls which type of value should be accepted/outputed (but might be wrong, this is just by reading the code).

About the custom fields and particularly the slider version: it will never work correctly as the JS and the HTML have both IDs hardcoded and you can't really have one ID for multiple fields, will never work (not something you did here and also I had criticised that when this field was introduced but as usual I was ignored)

**Comment by laoneo** [comment]
Would be great if you can help me out here.

> but as usual I was ignored

Your removed jQuery from core, what did you expect then :rofl:

**Comment by dgrammatiko** [comment]
> Your removed jQuery from core, what did you expect then

And that's my punishment for keeping the project somewhat competitive, fair enough 🙄

Joke aside, I'm kinda busy today, I'll try to sneak some time this weekend, unless someone is faster

**Pull request closed without merging**